File: fundamental/mkTriangle.hs
module Main where
import SAT.DIMACS
import System.Environment (getArgs)
import System.IO (hFlush, stdout)

chain :: Int -> [[Int]]
chain 1 = [[1]]
chain n = ([-(n-1) .. -1] ++ [n]) : chain (n -1)

main = do
  (n:_) <- getArgs
  toFile ("triangle-" ++ n ++ ".cnf") $ chain (read n)
  

File: fundamental/mkTriangle.rs
fn main() {
    make(2_000_000, 10);
}

fn make(nv: u32, chain: u32) {
    println!("p cnf {} {}", nv, nv);
    for i in 0..=nv - chain {
        print!("{} ", i + chain);
        for j in 1..chain {
            print!("-{} ", i + chain - j);
        }
        println!("0");
    }
    for i in 1..chain {
        println!("{} 0", i);
    }
}
